browser/search_engines/android/java/src/org/chromium/brave/browser/search_engines/settings/CustomSearchEnginesUtil.java

```diff
@@ -0,0 +1,173 @@
+        } catch (Exception e) {
+            Log.e(TAG, "Error saving search engines", e);
+        }
+    }
+
+    public static void addCustomSearchEngine(String searchEngineKeyword) {
+        List<String> customSearchEnginesList = getCustomSearchEngines();
+        if (customSearchEnginesList.isEmpty()) {
+            customSearchEnginesList = new LinkedList<>();
+        }
+        if (!customSearchEnginesList.contains(searchEngineKeyword)) {
+            customSearchEnginesList.add(searchEngineKeyword);
+            saveCustomSearchEngines(customSearchEnginesList);
+        }
+    }
+
+    public static List<String> getCustomSearchEngines() {
+        List<String> customSearchEnginesList = new LinkedList();
+        String savedSearchEngines =
+                ChromeSharedPreferences.getInstance().readString(CUSTOM_SEARCH_ENGINES, null);
+
+        if (savedSearchEngines == null) {
+            return customSearchEnginesList;
+        }
+
+        try {
+            // Wrap the array in an object to make it valid JSON
+            JSONObject wrapper =
+                    new JSONObject("{\"customSearchEngines\":" + savedSearchEngines + "}");
+            JSONArray searchEnginesArray = wrapper.getJSONArray("customSearchEngines");
+
+            for (int i = 0; i < searchEnginesArray.length(); i++) {
+                customSearchEnginesList.add(searchEnginesArray.getString(i));
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        return customSearchEnginesList;
+    }
+
+    public static boolean isCustomSearchEngineAdded(String searchEngineKeyword) {
+        List<String> customSearchEnginesList = getCustomSearchEngines();
+        return !customSearchEnginesList.isEmpty()
+                && customSearchEnginesList.contains(searchEngineKeyword);
+    }
+
+    public static void removeCustomSearchEngine(String searchEngineKeyword) {
+        List<String> customSearchEnginesList = getCustomSearchEngines();
+        if (!customSearchEnginesList.isEmpty()
+                && customSearchEnginesList.contains(searchEngineKeyword)) {
+            customSearchEnginesList.remove(searchEngineKeyword);
+            saveCustomSearchEngines(customSearchEnginesList);
+        }
+    }
+
+    public static void updateCustomSearchEngine(
+            String searchEngineKeyword, String newSearchEngineKeyword) {
+        List<String> customSearchEnginesList = getCustomSearchEngines();
+        if (!customSearchEnginesList.isEmpty()
+                && customSearchEnginesList.contains(searchEngineKeyword)) {
+            int index = customSearchEnginesList.indexOf(searchEngineKeyword);
+            if (index != -1) {
+                customSearchEnginesList.set(index, newSearchEngineKeyword);
+            }
+            saveCustomSearchEngines(customSearchEnginesList);
+        }
+    }
+
+    public static boolean isSearchQuery(String text) {
+        return text.contains("%s");
```

Comment:
I'm confused, the existence of `%s` is a search query for what? What text are you even passing in here? Also none of these methods appear to have any tests

Reply:
With this function, we check if `%s` is included in the url user passes. similar to what we are doing with IOS.

Reply:
We have to add a check for valid url with correct query as Template url service allows anything to be added. template url service doesn't seem to have any checks for the validation of the url.